test/unit/alignment/matrix/detail/simulated_alignment_test_template.hpp

```diff
@@ -0,0 +1,140 @@
+// -----------------------------------------------------------------------------------------------------
+// Copyright (c) 2006-2019, Knut Reinert & Freie Universität Berlin
+// Copyright (c) 2016-2019, Knut Reinert & MPI für molekulare Genetik
+// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
+// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE.md
+// -----------------------------------------------------------------------------------------------------
+
+#include <gtest/gtest.h>
+
+#include <seqan3/std/algorithm>
+
+template <typename t>
+struct simulated_alignment_test : public ::testing::Test
+{
+    using score_type = typename t::score_type;
+
+    score_type match = 0;
+    score_type mismatch = -1;
+    score_type gap = -1;
+
+    std::string first{"abba"};
+    std::string second{"baba"};
+
+    t test_data{first, second};
+
+    auto & matrix()
+    {
+        return test_data.matrix;
+    };
+
+    auto & gold_matrix()
+    {
+        return test_data.gold_matrix;
+    }
+
+    auto last_init_column()
+    {
+        return test_data.last_init_column;
+    }
+};
+
+TYPED_TEST_CASE_P(simulated_alignment_test);
+
+using namespace seqan3;
+using namespace seqan3::detail;
+
+TYPED_TEST_P(simulated_alignment_test, linear_alignment)
+{
+    std::vector<typename TypeParam::score_type> cmp_matrix{};
+    cmp_matrix.resize(this->gold_matrix().size());
+    auto cmp_it = cmp_matrix.begin();
+
+    // ----------------------------------------------------------------------------
+    // Initialise the first column
+    // ----------------------------------------------------------------------------
+
+    auto mat_it = this->matrix().begin();
+    auto col = *mat_it;
+    auto col_it = col.begin();;
+    (*col_it).current = 0;
+    (*col_it).up = this->gap;
+    (*col_it).w_left = this->gap;
```

Comment:
Why don't you use `->`operator?

Reply:
because it can't. The ->operator does not work for the iterator since it returns a temporary. It cannot return a pointer to a temporary then.